### PR TITLE
Add allowed_mapgens option in game.conf.

### PR DESCRIPTION
**Pull request by wt**
The game.conf has a disallowed_mapgens option. However, some games
require a certain mapgen to be used, like the CTF plugin. This change
adds an option to specify allowed mapgens so that the setting can be
specified in a way that needn't be updated as map generators are added
to minetest.

This change doens't address any outstanding issue that I found on github.
I found the issue while my son was trying to play with the CTF mod. When
I finally realized that the problem was that the wrong mapgen was used, this
idea came to me. I hope you find it useful.

## Status

This PR is Ready for Review.

## How to test

* Add the following to games/minimal/game.conf:

    allowed_mapgens = v7,v6

* Start minetest and create a new map for the minimal game. You should only
  be given the v6 and v7 options for mapgens.

### CODE REVIEW

**Comment by paramat** [comment]
Seems useful.

**Comment by wt** [comment]
I think it's ready to merge if y'all want. I want to add the field to the ctf game after this lands.

**Comment by Wuzzy2** [comment]
What happens when used in combination with `disallowed_mapgens`? The documentation doesn't say, therefore :-1: for now.

**Comment by wt** [comment]
Practically speaking, both settings can exist in a single game. In that case, the allowed is applied before the disallowed. For example, if you have the following config:

    allowed_mapgens = v7
    disalllowed_mapgens = v6

You will get only v7 as allowed. If you put the same in both, that's where you have a problem. Here's an example.

    allowed_mapgens = v7
    disallowed_mapgens = v7

Now, no mapgens will be available, and the drop down in the ui will be blank, which also won't let you select one and thus you can't create a new map.

I think there are valid use cases for including both however. Consider the following:

    allowed_mapgens = singlenode
    disallowed_mapgens = v6,v7,...all others that aren't singlenode...

This would allow a game like ctf to define a disallowed mapgens so that it is can provide a better experience on the current version of minetest while only allowing singlenode in the future. The "disallowed_mapgens" wouldn't need to be updated if a mapgen is added to a future version of minetest. And this would provide a much better experience than what you get today, which is to create a non-singlenode map and then get an assertion when trying to start the game on that map.

Would it be enough to document this behavior explicitly in order to get this change merged?

**Comment by SmallJoker** [comment]
@Wuzzy2 The settings could be stated to be incompatible. Game writers want both options, but will only use one setting: (examples)
1) minimal -> `disallowed_mapgens = singlenode` (needs terrain)
2) ctf -> `allowed_mapgens = singlenode` (only singlenode is OK)

It'll be up to the game writers to decide how to handle future mapgens.

**Comment by Wuzzy2** [comment]
I don't really care about how it behaves tbh, I just want that, whatever happens, should be documented. @wt's explanation could be summarized in maybe 1-2 sentences.

**Comment by wt** [comment]
I'll take a crack at documenting it when I get home. Do you prefer me to amend the commit and rebase or just stack commits?

**Comment by SmallJoker** [comment]
@wt This PR is small enough so that it does not matter what you do with the branch as long it's based on master without foreign commits. If there are multiple commits, they can be squashed & edited upon merge by a few clicks. FYI

Offtopic information.

**Comment by wt** [comment]
I updated the docs.

File: doc/lua_api.txt [hunk 1]
```diff
@@ -68,9 +68,12 @@ The game directory can contain the following files:
       e.g. `allowed_mapgens = v5,v6,flat`
       Mapgens not in this list are removed from the list of mapgens for
       the game.
+      If not specified, all mapgens are allowed.
     * `disallowed_mapgens = <comma-separated mapgens>`
       e.g. `disallowed_mapgens = v5,v6,flat`
       These mapgens are removed from the list of mapgens for the game.
+      When both `allowed_mapgens` and `disallowed_mapgens` are both
+      specified, `allowed` is applied before `disallowed'.
```
Comment (SmallJoker) [inline doc/lua_api.txt]: First "both" is redundant. Code tags should contain code or complete setting names.

Suggestion: `allowed_mapgens` is applied first when used in combination.

Comment (wt) [reply doc/lua_api.txt]: @SmallJoker, Thanks for that feedback. I update dropped a new commit. How do you feel about this one?

**Comment by wt** [comment]
I see an approval, but no merge. Is there anything else I need to do to get this merged?

**Comment by SmallJoker** [comment]
@wt A second approval or a Trivial tag.

**Comment by wt** [comment]
Hopefully someone will TAL at some point. Thanks for your help.

**Comment by wt** [comment]
@paramat, is this useful enough to get your approval?

**Comment by paramat** [comment]
I just came here to say i will probably approve this ince i get time to review it =D

**Comment by wt** [comment]
Any reviewers around? This is a pretty tiny change that I think would improve users' lives a little bit.

**Comment by paramat** [comment]
Sorry, i am very busy and am prioritising.

**Comment by wt** [comment]
Any chance anyone has time for a pretty simple review? I'd really love to see this get in.

**Comment by paramat** [comment]
It will probably be merged in 5.3.0-dev. We are in feature freeze for release of 5.2.0 so there is no point reviewing it now, and it will not be merged for 5.2.0. This is partly why this is not getting attention currently, we have to prioritise releasing.

**Pull request merged**